controller/src/beerocks/master/son_master_thread.cpp

```diff
@@ -1186,16 +1191,162 @@ bool master_thread::handle_cmdu_1905_higher_layer_data_message(Socket *sd,
     return son_actions::send_cmdu_to_agent(sd, cmdu_tx);
 }
 
+bool master_thread::handle_cmdu_1905_topology_notification(Socket *sd,
+                                                           ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received TOPOLOGY_NOTIFICATION_MESSAGE, mid=" << std::hex << int(mid);
+
+    std::shared_ptr<wfa_map::tlvClientAssociationEvent> client_association_event_tlv = nullptr;
+    std::shared_ptr<beerocks_message::tlvVsClientAssociationEvent> vs_tlv            = nullptr;
+
+    int type;
+    while ((type = cmdu_rx.getNextTlvType()) != int(ieee1905_1::eTlvType::TLV_END_OF_MESSAGE)) {
+        switch (type) {
+        case int(wfa_map::eTlvTypeMap::TLV_CLIENT_ASSOCIATION_EVENT):
+            LOG(DEBUG) << "Found TLV_CLIENT_ASSOCIATION_EVENT TLV";
+            client_association_event_tlv = cmdu_rx.addClass<wfa_map::tlvClientAssociationEvent>();
+            if (!client_association_event_tlv) {
+                LOG(ERROR) << "addClass wfa_map::tlvClientAssociationEvent failed";
+                return false;
+            }
+            break;
+        case int(ieee1905_1::eTlvType::TLV_VENDOR_SPECIFIC):
+            LOG(DEBUG) << "Found TLV_VENDOR_SPECIFIC TLV";
+
+            if (!message_com::parse_intel_vs_message(cmdu_rx)) {
+                LOG(DEBUG) << "Failed to parse intel vs message (not Intel?)";
+                continue;
+            }
+
+            vs_tlv = cmdu_rx.addClass<beerocks_message::tlvVsClientAssociationEvent>();
+            if (!vs_tlv) {
+                LOG(ERROR) << "addClass wfa_map::tlvVsClientAssociationEvent failed";
+                return false;
+            }
+            break;
+        default:
+            LOG(ERROR) << "Unexpected TLV. Ignore CMDU.";
+            return false;
+        }
+    }
+
+    if (!client_association_event_tlv) {
+        LOG(ERROR) << "wfa_map::tlvClientAssociationEvent not found";
+        return false;
+    }
+
+    auto &client_mac    = client_association_event_tlv->client_mac();
```

Comment:
My thumb rule is that if a parameter size is greater than sizeof(int) which is the size of a pointer, use reference. I'm not sure what the compiler does, and I don't think anyone can know without actually peeking assembly code, especially when the compiler may be changed. So using reference seems to be the right decision for me. 
Regarding the "const", I disagree it is necessary. At this place, I don't mind if it will be changed and it does not create security issue like when not using const on function argument.

Since association_event is one byte, I will change it to copy.

Reply:
> My thumb rule is that if a parameter size is greater than sizeof(int) which is the size of a pointer, use reference. I'm not sure what the compiler does, and I don't think anyone can know without actually peeking assembly code, especially when the compiler may be changed. So using reference seems to be the right decision for me.

[Stroustroup disagrees with you](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rf-in): "up to two or three words".

> Regarding the "const", I disagree it is necessary. At this place, I don't mind if it will be changed and it does not create security issues like when not using const on function argument.

Indeed, since both `cmdu_rx` and `client_association_event_tlv` are non-const, there's no reason to make `client_mac` const.

Personally, I would prefer all of them to be const. But that's not possible at the moment because addClass modifies the object.

Reply:
> > My thumb rule is that if a parameter size is greater than sizeof(int) which is the size of a pointer, use reference. I'm not sure what the compiler does, and I don't think anyone can know without actually peeking assembly code, especially when the compiler may be changed. So using reference seems to be the right decision for me.
> 
> [Stroustroup disagrees with you](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rf-in): "up to two or three words".

BTW this pattern is used in many other places in the code even for small objects, so there's no need to change it. I.e. (nitpick)